microsoft/PowerToys
# #19638 Added the option to Mouse Crosshairs to fix the lengths of the…

… crosshairs. Some discussion about design in the issue but currently implemented as a configuration toggle and separate length input.


## Summary of the Pull Request
Adds to the Mouse Crosshairs utility the ability for users to configure fixed-length mouse crosshairs rather than always having the crosshairs extend to the edge of the screen.  If fixed length toggle is enabled, users may configure a crosshairs length of at least 1 pixel.


## PR Checklist

- [x] **Closes:** #19638
- [No existing test infrastructure] **Tests:** Added/updated and all pass
- [Believe so] **Localization:** All end user facing strings can be localized
- [Need direction if there's any to update] **Dev docs:** Added/updated
- [I don't believe there's any - please let me know if I need to complete any below/additional steps] **New binaries:** Added on the required places
- [Need direction if there's any to update] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx


## Detailed Description of the Pull Request / Additional comments
Added Mouse Crosshairs configuration toggle and fixed length input text box that is enabled if the toggle is enabled.  I do not know what verbiage or configuration conventions/idioms the PowerToys team prefers, I didn't see any clear guidelines in the docs (please let me know if I missed them!), and I wasn't able to get a clear answer in the issue in the last 2 weeks, so I am completely open to reworking the configuration UX however the team prefers.

Adjusted crosshairs drawing logic to check if fixed length is enabled and draw accordingly, which involved extending the crosshair border beyond the far side when using a fixed length.  I also consolidated the crosshairs drawing logic/calculation a bit as expressions were repeated in several places.  I chose to use inline scopes to ensure no temporary variables from one crosshairs component leaked into others; I am, of course, open to changing it if this violates some policy (I didn't see one).

Note: For screen edge interaction, I felt it was most intuitive to have the crosshairs extend onto adjacent monitors in fixed-length mode.  This makes the locating the crosshairs during the transition between monitors easier- instead of popping onto the next monitor, it transitions smoothly (or as smoothly as can be expected, in the case of different-resolution monitors).


## Validation Steps Performed
Did not see any existing automated testing.

Enabled toggle and configured various fixed lengths with various crosshairs borders configurations.  Ensured the toggle was not enabled by default when transitioning between the current version of PowerToys and a version containing these updates.  Checked interaction with all 4 edges of the screen using a multi-monitor setup and configuring Windows monitors accordingly.
